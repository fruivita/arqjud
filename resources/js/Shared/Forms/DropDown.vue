<!--
    Componente Select com Ã­cone (opcional).

    @see https://vuejs.org/guide/introduction.html
    @see https://tailwindcss.com/docs
    @see https://inertiajs.com/
 -->

<script>
export default { inheritAttrs: false };
</script>

<script setup>
import { gerarID } from '@/Composables/UseGerarID';
import MensagemErro from '@/Shared/Forms/MensagemErro.vue';
import Icone from '@/Shared/Icones/Icone.vue';
import { useTranslationsStore } from '@/Stores/TranslationsStore';
import { get, isArray, join, pick, values } from 'lodash';
import { computed } from 'vue';

const props = defineProps({
    opcoes: { type: Array, required: true },
    id: { type: String },
    icone: { type: String },
    erro: { type: String },
    label: { type: String },
    modelValue: { type: [Number, String] },
    labelOpcao: { type: [Array, String], default: 'nome' },
    disabled: { type: Boolean, default: false },
    required: { type: Boolean, default: false },
});

const __ = useTranslationsStore().__;

const emit = defineEmits(['update:modelValue']);

const idReal = computed(() => props.id ?? gerarID());

const optionText = (opcao) => {
    let texto = '';

    if (isArray(props.labelOpcao)) {
        texto = join(values(pick(opcao, props.labelOpcao)), '/');
    } else {
        texto = get(opcao, props.labelOpcao);
    }

    return texto;
};

const onChange = (event) => {
    if (!props.disabled) {
        emit('update:modelValue', event.target.value);
    }
};
</script>

<template>
    <div class="w-full space-y-1">
        <label v-if="label" :for="idReal" class="font-bold" dusk="label">
            {{ label }}

            <span v-if="required" class="text-red-500" dusk="required">*</span>
        </label>

        <div
            :class="{
                'dark:text-primaria-900': !disabled,
                'dark:text-primaria-50': disabled,
            }"
            class="relative"
        >
            <div
                v-if="icone"
                class="pointer-events-none absolute inset-y-0 left-0 z-10 flex items-center pl-3"
            >
                <Icone :nome="icone" />
            </div>

            <select
                :id="idReal"
                :class="{
                    'pl-12': icone,
                    'pl-2': !icone,
                    'border-primaria-500 bg-primaria-100 ring-primaria-500': !erro,
                    'border-red-500 bg-red-100 ring-red-500': erro,
                    'border-dashed': disabled,
                }"
                :disabled="disabled"
                :required="required"
                :value="modelValue"
                @change="onChange"
                v-bind="$attrs"
                class="w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900"
            >
                <template v-if="opcoes.length">
                    <option v-for="opcao in opcoes" :key="opcao.id" :value="opcao.id">
                        {{ optionText(opcao) }}
                    </option>
                </template>

                <template v-else>
                    <option value="-1" selected>{{ __('Nenhum registro encontrado') }}</option>
                </template>
            </select>
        </div>

        <MensagemErro v-if="erro" :erro="erro" />
    </div>
</template>
