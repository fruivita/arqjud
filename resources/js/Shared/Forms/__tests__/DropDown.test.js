/**
 * Testes para o componente DropDown.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import DropDown from '@/Shared/Forms/DropDown.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test, vi } from 'vitest';

vi.mock('@/Composables/UseGerarID', () => ({
    __esModule: true,
    ...vi.importActual('@/Composables/UseGerarID'),
    gerarID: () => 'foo-id',
}));

let mountFunction;

beforeEach(() => {
    const pinia = () => createTestingPinia({ stubActions: false });

    mountFunction = (options = {}) => {
        return mount(DropDown, {
            global: {
                plugins: [pinia()],
            },
            ...options,
        });
    };
});

describe('DropDown', () => {
    test('propriedades estão definidas', () => {
        expect(DropDown.props).toMatchObject({
            opcoes: { type: Array, required: true },
            id: { type: String },
            icone: { type: String },
            erro: { type: String },
            label: { type: String },
            modelValue: { type: [Number, String] },
            labelOpcao: { type: [Array, String], default: 'nome' },
            disabled: { type: Boolean, default: false },
            required: { type: Boolean, default: false },
        });
    });

    test('renderiza o label respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    label: 'loren',
                    modelValue: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o símbolo de campo obrigatório respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    label: 'loren',
                    modelValue: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                    required: true,
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o ícone respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    icone: 'key',
                    modelValue: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza a mensagem de erro respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    modelValue: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                    erro: 'loren',
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza a propriedade escolhida do objeto respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    modelValue: undefined,
                    labelOpcao: 'extra',
                    opcoes: [
                        { id: 1, nome: 'bar', extra: 'loren' },
                        { id: 2, nome: 'foo', extra: 'ipsun' },
                    ],
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza múltiplas propriedades do objeto respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    modelValue: undefined,
                    opcoes: [
                        { id: 1, extra: 'bar', outra: 'loren' },
                        { id: 2, extra: 'foo', outra: 'ipsun' },
                    ],
                    labelOpcao: ['extra', 'outra'],
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('atributos do componente e extras renderizados corretamente respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    modelValue: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                    extra: 'loren',
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('rendiza o componente de maneira diversa quando desabilitado respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    modelValue: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                    disabled: true,
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('rendiza o componente quando não há opções para o select respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    modelValue: undefined,
                    opcoes: [],
                    disabled: true,
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('atualiza o v-model ao alterar o valor do elemento', async () => {
        const parent = mount({
            data: () => {
                return {
                    selected: 2,
                    opcoes: [
                        { id: 1, nome: 'bar' },
                        { id: 2, nome: 'foo' },
                    ],
                };
            },
            template:
                '<div> <DropDown v-model="selected" :opcoes="opcoes" :maxlength="10"/> </div>',
            components: { DropDown: DropDown },
        });

        const dropdown = parent.find('select');

        dropdown.element.value = 1;
        await dropdown.trigger('change');

        expect(parent.vm.selected).toBe('1');
    });

    test('dispara o evento onChange se o componente estiver habilitado', () => {
        const wrapper = mountFunction({
            props: {
                modelValue: 2,
                maxlength: 10,
                opcoes: [
                    { id: 1, nome: 'bar' },
                    { id: 2, nome: 'foo' },
                ],
            },
        });

        const event = { target: { value: 1 } };

        wrapper.vm.onChange(event);

        expect(wrapper.emitted()['update:modelValue'][0]).toEqual([1]);
    });

    test('não dispara o evento onChange se o componente estiver desabilitado', () => {
        const wrapper = mountFunction({
            props: {
                modelValue: 2,
                maxlength: 10,
                opcoes: [
                    { id: 1, nome: 'bar' },
                    { id: 2, nome: 'foo' },
                ],
                disabled: true,
            },
        });

        const event = { target: { checked: true } };

        wrapper.vm.onChange(event);

        expect(wrapper.emitted()['update:modelValue']).toBeFalsy();
    });
});
