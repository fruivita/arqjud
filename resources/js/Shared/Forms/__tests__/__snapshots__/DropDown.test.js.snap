// Vitest Snapshot v1

exports[`DropDown > atributos do componente e extras renderizados corretamente respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-900 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\" extra=\\"loren\\">
      <option value=\\"1\\">bar</option>
      <option value=\\"2\\">foo</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza a mensagem de erro respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-900 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-red-500 bg-red-100 ring-red-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\">
      <option value=\\"1\\">bar</option>
      <option value=\\"2\\">foo</option>
    </select>
  </div>
  <p class=\\"text-sm font-bold text-red-500\\" dusk=\\"erro\\">loren</p>
</div>"
`;

exports[`DropDown > renderiza a propriedade escolhida do objeto respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-900 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\">
      <option value=\\"1\\">loren</option>
      <option value=\\"2\\">ipsun</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza múltiplas propriedades do objeto respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-900 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\">
      <option value=\\"1\\">bar/loren</option>
      <option value=\\"2\\">foo/ipsun</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza o componente de maneira diversa quando desabilitado respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-50 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 border-dashed w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\" disabled=\\"\\">
      <option value=\\"1\\">bar</option>
      <option value=\\"2\\">foo</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza o componente quando não há opções para o select respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-50 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 border-dashed w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\" disabled=\\"\\">
      <option value=\\"-1\\">Nenhum registro encontrado</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza o label respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\"><label for=\\"foo-id\\" class=\\"font-bold\\" dusk=\\"label\\">loren
    <!--v-if-->
  </label>
  <div class=\\"dark:text-primaria-900 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\">
      <option value=\\"1\\">bar</option>
      <option value=\\"2\\">foo</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza o símbolo de campo obrigatório respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\"><label for=\\"foo-id\\" class=\\"font-bold\\" dusk=\\"label\\">loren <span class=\\"text-red-500\\" dusk=\\"required\\">*</span></label>
  <div class=\\"dark:text-primaria-900 relative\\">
    <!--v-if--><select id=\\"foo-id\\" class=\\"pl-2 border-primaria-500 bg-primaria-100 ring-primaria-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\" required=\\"\\">
      <option value=\\"1\\">bar</option>
      <option value=\\"2\\">foo</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDown > renderiza o ícone respeitando o snapshot 1`] = `
"<div class=\\"w-full space-y-1\\">
  <!--v-if-->
  <div class=\\"dark:text-primaria-900 relative\\">
    <div class=\\"pointer-events-none absolute inset-y-0 left-0 z-10 flex items-center pl-3\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
        <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
      </svg></div><select id=\\"foo-id\\" class=\\"pl-12 border-primaria-500 bg-primaria-100 ring-primaria-500 w-full rounded-lg border py-2 opacity-100 outline-none transition duration-300 focus:ring disabled:cursor-not-allowed disabled:bg-primaria-50 dark:disabled:bg-secundaria-900\\">
      <option value=\\"1\\">bar</option>
      <option value=\\"2\\">foo</option>
    </select>
  </div>
  <!--v-if-->
</div>"
`;
